mmda/utils/tools.py

```diff
@@ -66,4 +20,116 @@ def allocate_overlapping_tokens_for_box(
+
+    def build_graph_index_overlap(self):
+        """
+        Build graph, each node is represented by (start, end) of tuple, with the list of spans with overlapping
+        boxes given, w, h
+        """
+        is_neighboring_spans = (
+            lambda span1, span2: min(
+                abs(span1.start - span2.end), abs(span1.end - span2.start)
+            ) <= self.distance
+        )
+
+        for i, span_i in enumerate(self.list_of_spans):
+            for j in range(i + 1, len(self.list_of_spans)):
+                if is_neighboring_spans(span_i, self.list_of_spans[j]):
+                    self.graph[span_i.start, span_i.end].append(self.list_of_spans[j])
+                    self.graph[self.list_of_spans[j].start, self.list_of_spans[j].end].append(span_i)
+
+    def build_graph_box_overlap(self):
+        """
+        Build graph, each node is represented by (start, end) of tuple, with the list of spans with overlapping
+        boxes given, w, h
+        """
+        for i, span_i in enumerate(self.list_of_spans):
+            assert hasattr(span_i, 'box'), 'Missing attribute box in a span'
+            for j in range(i + 1, len(self.list_of_spans)):
+                assert hasattr(self.list_of_spans[j], 'box'), 'Missing attribute box in a span'
+                if span_i.box.is_overlap(self.list_of_spans[j].box, self.w, self.h):
+                    self.graph[span_i.start, span_i.end].append(self.list_of_spans[j])
+                    self.graph[self.list_of_spans[j].start, self.list_of_spans[j].end].append(span_i)
+
+    # gets the connected components of the boxes overlap graph.
+    def get_components(self):
+        """
+        Groups connected graph nodes into dictionary list
+        """
+        visited = set()
+        comp_number = 0
+        nodes_in_comp = defaultdict(list)
+
+        def mark_component_dfs(start):
+            stack = [start]
+            while stack:
+                span = stack.pop()
+                node = span.start, span.end
+                if node not in visited:
+                    visited.add(node)
+                    nodes_in_comp[comp_number].append(span)
+                    stack.extend(self.graph[node])
+
+        # mark all nodes in the same connected component with the same integer.
+        for span in self.list_of_spans:
+            center = span.start, span.end
+            if center not in visited:
+                mark_component_dfs(span)
+                comp_number += 1
+
+        return nodes_in_comp, comp_number
+
+    def merge_neighbor_spans_by_box_coordinate(self):
+        """
+        For each of the lists of the connected nodes determined by distnace between the boxes,
+        merge boxes and find, min, max of the index
+        """
+        self.build_graph_box_overlap()
+        nodes_in_comp, number_of_comps = self.get_components()
+
+        # all intervals in each connected component must be merged.
+        merged_spans = []
```

Comment:
can you encapsulate L111-L117 and L128-L134 into a single method called `build_merged_spans_from_connected_components()`?  it looks like these are identical between the two methods